libbeat/template/load.go

```diff
@@ -66,67 +75,92 @@ func NewLoader(cfg *common.Config, client ESClient, beatInfo beat.Info, fields [
+			return fmt.Errorf("error creating template instance: %v", err)
+		}
 
-		logp.Info("Loading template for Elasticsearch version: %s", l.client.GetVersion())
-		if l.config.Overwrite {
-			logp.Info("Existing template will be overwritten, as overwrite is enabled.")
+		templateName := tmpl.GetName()
+		if cfg.JSON.Enabled {
+			templateName = cfg.JSON.Name
 		}
+		// Check if template already exist or should be overwritten
+		loaded := l.templateLoaded(templateName)
+		if !loaded || cfg.Overwrite {
 
-		var template map[string]interface{}
-		if l.config.JSON.Enabled {
-			jsonPath := paths.Resolve(paths.Config, l.config.JSON.Path)
-			if _, err := os.Stat(jsonPath); err != nil {
-				return fmt.Errorf("error checking for json template: %s", err)
+			logp.Info("Loading template for Elasticsearch version: %s", l.esVersion)
+			if cfg.Overwrite {
+				logp.Info("Existing template will be overwritten, as overwrite is enabled.")
 			}
 
-			logp.Info("Loading json template from file %s", jsonPath)
-
-			content, err := ioutil.ReadFile(jsonPath)
-			if err != nil {
-				return fmt.Errorf("error reading file. Path: %s, Error: %s", jsonPath, err)
-
-			}
-			err = json.Unmarshal(content, &template)
-			if err != nil {
-				return fmt.Errorf("could not unmarshal json template: %s", err)
+			var template common.MapStr
+			if cfg.JSON.Enabled {
+				jsonPath := paths.Resolve(paths.Config, cfg.JSON.Path)
+				if _, err := os.Stat(jsonPath); err != nil {
+					return fmt.Errorf("error checking for json template: %s", err)
+				}
+
+				logp.Info("Loading json template from file %s", jsonPath)
+
+				content, err := ioutil.ReadFile(jsonPath)
+				if err != nil {
+					return fmt.Errorf("error reading file. Path: %s, Error: %s", jsonPath, err)
+
+				}
+				err = json.Unmarshal(content, &template)
+				if err != nil {
+					return fmt.Errorf("could not unmarshal json template: %s", err)
+				}
+				// Load fields from path
+			} else if cfg.Fields != "" {
+				logp.Debug("template", "Load fields.yml from file: %s", cfg.Fields)
+
+				fieldsPath := paths.Resolve(paths.Config, cfg.Fields)
+
+				template, err = tmpl.LoadFile(fieldsPath)
+				if err != nil {
+					return fmt.Errorf("error creating template from file %s: %v", fieldsPath, err)
+				}
+
+				// Load fields for modules
+			} else if cfg.Modules != "" {
+				logp.Debug("template", "Load fields for %s", cfg.Modules)
+
+				fields, err := asset.GetFieldsFor(l.beatInfo.Beat, strings.Split(cfg.Modules, ","))
+				if err != nil {
+					return err
+				}
+				template, err = tmpl.LoadBytes(fields)
+				if err != nil {
+					return fmt.Errorf("error creating template: %v", err)
+				}
+
+				// Load default fields
+			} else {
+				logp.Debug("template", "Load default fields")
+				fields, err := asset.GetFields(l.beatInfo.Beat)
+				if err != nil {
+					return err
+				}
+				template, err = tmpl.LoadBytes(fields)
+				if err != nil {
+					return fmt.Errorf("error creating template: %v", err)
+				}
 			}
```

Comment:
We could probably split that method with different format concerns?

Reply:
Can do that yes, haven't invested into refactoring yet.